src/main/java/com/cloudbees/plugins/deployer/engines/Engine.java

```diff
@@ -232,7 +234,7 @@ public void log(String message) {
         throw new DeployException("Deployment hosts of type " + configuration.getClass() + " are unsupported");
     }
 
-    public static class FingerprintDecorator implements Callable<DeployedApplicationLocation, IOException> {
+    public static class FingerprintDecorator extends MasterToSlaveCallable<DeployedApplicationLocation, IOException> {
```

Comment:
I do not think this is true. `FingerprintingWrapper` has code which apparently calls it from the agent side, if _that_ callable was running remotely. And its body calls `Hudson.getInstance()` without a null check, confirming that it is intended to be run on the master side. Thus it should be a `SlaveToMasterCallable`.

But before you make such a change you would need to review its implementation carefully to ensure that no malicious arguments could possibly cause any harm. That seems unlikely, since the agent could pass any `md5sum` it likes, mangling fingerprints of some unrelated artifact. Which means that really `FingerprintDecorator` needs to be deleted, and `FingerprintingWrapper` needs to be reworked so that any _information_ about fingerprints is simply `return`ed, to be handled by its caller (after any necessary sanity checks). That in turn means understanding the call points in the two `Engine.process` overloads. The one taking `FilePath` is called from `perform` in what is clearly a master context (because it is calling `build.getWorkspace()`, and a `Run` can only exist on master); thus the other overload must also be called from master. Thus it is the `process` methods which would need to add suitable `DeployedApplicationFingerprintFacet`s; perhaps `FingerprintingWrapper` needs to be returning a pair of `DeployedApplicationLocation` and `md5sum`.